snuba/utils/streams/consumer.py

```diff
@@ -97,6 +97,30 @@ def unsubscribe(self) -> None:
     def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
         raise NotImplementedError
 
+    @abstractmethod
+    def pause(self, partitions: Sequence[Partition]) -> None:
+        """
+        Pause consuming from the provided partitions.
+
+        A partition that is paused will be automatically resumed during
+        reassignment. If partitions should remain paused across rebalances,
```

Comment:
what was the reason for this choice ? Is it something imposed by the confluent api ? Or is it a design choice to ensure the client would not try to resume a set of partitions that does not correspond anymore to the set of partitions assigned to the consumer ?

Reply:
This prevents this case from happening:

1. some state change occurs that causes a partition in the assignment set to be paused (e.g. remote offset aligns with local offset in the synchronized consumer case)
2. the assignment is revoked, and whatever state that led to that partition being paused is removed from local memory
3. an unbounded amount of time goes by, consumer deals with zero or more other assignments that do not include the partition w were talking about previously
4. the partition reappears in the assignment set, at this point state is carried over from the previous assignment (pause/resume status) but the state that caused the partition to be paused in the first place may not be

I think it is reasonable from the consumer perspective to consider each assignment as a clean slate that should not be affected by any prior consumer state other than what is tracked by the broker (pause/resume is local to the consumer itself, if that was not clear already) — *unless* it is specifically handled by the caller of `subscribe` in the assignment callback.

This also makes the behavior consistent with what would happen if:

1. the assignment was relocated from consumer A to consumer B: consumer B will have no notion that consumer A had previously paused the partition, so it would start consuming the partition (since not paused is the default state.)
2. the consumer was to crash and restart with an identical assignment set (really just another form of rebalancing)

Reply:
The last two points convinced me. I don;t think there is a way to recover the state the assignment was into.
I would consider adding them to the comment so that someone does not decide to change this behavior later on.